### ISSUE TITLE
[SuperEditor][SuperReader] Improve accessibility / semantics

### ISSUE THREAD

**Issue by Jethro87** [superlistapp/super_editor#1072]
We have users that use accessibility tools (such as screen readers) to edit/read their documents. Since migrating to super_editor, we've had reports that screen readers no longer work.

According to our users:
- When the doc does not have selection, screen readers read nothing.
- When the doc has selection (but the user is not currently typing), screen readers can read the document.
- When the user is typing/editing the document, screen readers stop working.
-  Users can't scroll/move the carat with a screen reader anymore ("I can't scroll by characters or words to correct any errors because my screenreader just doesn't work well with documents anymore").

I did some testing with the example editor and Accessibility Inspector with XCode/Simulator. Here's what I found:

### With no selection, only checkboxes are recognized by accessibility inspector
This is because the default Flutter `Checkbox` contains a `Semantic` widget:

```
    return Semantics(
      checked: widget.value ?? false,
      mixed: widget.tristate ? widget.value == null : null,
      child: buildToggleable(
...
```
<img width="367" alt="checkbox-super-editor" src="https://user-images.githubusercontent.com/6467808/227791418-da7b6d18-9449-4ce6-8251-a07cf7d506d0.png">

<img width="576" alt="checkbox-accessibility-inspector" src="https://user-images.githubusercontent.com/6467808/227791423-2e3cc9b5-627f-495f-a5e4-0df64ace0467.png">

### After modifying `ParagraphWithHintComponent` to contain a `Semantics` widget, it is recognized by accessibility inspector when selection is null

```
    Semantics(
      label: 'Paragraph',
      textField: true,
      multiline: true,
      value: componentViewModel.text.text,
      child: TextWithHintComponent(
        key: componentContext.componentKey,
        text: componentViewModel.text,
...
```

<img width="393" alt="paragraph-with-semantics-super-editor-no-selection" src="https://user-images.githubusercontent.com/6467808/227791513-1bd2f409-3458-4e7a-8926-e403e81c4c49.png">

<img width="574" alt="paragraph-accessibility-inspector-no-selection" src="https://user-images.githubusercontent.com/6467808/227791521-2dbabc36-a5ba-4057-b712-727c3e8f2d80.png">

### When the same `ParagraphWithHintComponent` has selection, it is not recognized by accessibility inspector

<img width="360" alt="paragraph-with-semantics-super-editor-with-selection" src="https://user-images.githubusercontent.com/6467808/227791762-12077e7f-6a4b-4484-b88d-4551c843456d.png">

<img width="574" alt="paragraph-accessibility-inspector-with-selection" src="https://user-images.githubusercontent.com/6467808/227791634-8a733920-b38b-4def-acae-34553e20a318.png">

---

It seems to me that, at the very least, all super_editor elements should be wrapped with `Semantic` widgets. I haven't done enough research to understand why, when the document has selection (and the modified `ParagraphWithHintComponent` is being used, wrapped with a `Semantics` widget), the semantics don't appear as they did when the document did not have selection. And why the user is not able to move the carat / modify the document with a screen reader.

**Comment by Jethro87** [comment]
cc @matthew-carroll

**Comment by matthew-carroll** [comment]
@Jethro87 do you think you could break this into individual tickets for each observable accessibility goal that you'd like to see?
